ir/src/ir2/graph/nodes.rs

```diff
@@ -0,0 +1,236 @@
+use crate::graph::{BackLink, Child, Graph, Leaf, Link, Node, Parent};
+use std::fmt::Debug;
+use std::ops::{Deref, DerefMut};
+
+#[derive(Clone, Eq, PartialEq)]
+pub enum NodeTypes {
+    Graph(Graph),
+    Node(Node),
+    I32(Leaf<i32>),
+    Add(Add),
+    Function(Function),
+}
+
+impl Debug for NodeTypes {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            NodeTypes::Graph(graph) => write!(f, "{:?}", graph),
+            NodeTypes::Node(node) => write!(f, "{:?}", node),
+            NodeTypes::I32(leaf) => write!(f, "{:?}", leaf),
+            NodeTypes::Add(add) => write!(f, "{:?}", add),
+            NodeTypes::Function(function) => write!(f, "{:?}", function),
+        }
+    }
+}
+
+impl Parent for Link<NodeTypes> {
+    /// Add a child to the node.and set the parent of the child to the node.
+    /// Returns the node.
+    fn add_child(&mut self, child: Link<NodeTypes>) -> Link<NodeTypes> {
+        println!(
+            "{:?}.add_child({:?}, {:?})",
+            std::any::type_name::<Self>(),
+            self,
+            child
+        );
+        let mut not_a_node = false;
+        match self.borrow_mut().deref_mut() {
+            NodeTypes::Graph(graph) => {
+                graph.add_child(child.clone());
+            }
+            NodeTypes::Node(node) => {
+                node.add_child(child.clone());
+            }
+            NodeTypes::Add(add) => {
+                add.add_child(child.clone());
+            }
+            _ => {
+                // a self field is mutably borrowed and dereferenced
+                // so self is not available for borrowing and debugging
+                not_a_node = true;
+            }
+        }
+        if not_a_node {
+            panic!("add_child called on non-node: {:?}", self);
```

Comment:
weird naming there, I think having an helper function `is_node` / `can_have_children` on NodeTypes is clearer. Then, once you check, you can have "unreachable".
and change the message to "on a NodeTypes that does not impl Parent" maybe

Reply:
Good call, thats probably a way to customize the node's behaviour in traits with default implementations too, via such helpers.